### PR TITLE
[node-core-library] Add AnsiEscape API

### LINKED ISSUE #2135: [heft] Rush stream-collator does not collate timing messages
## Please prefix the issue title with the project name i.e. [rush], [api-extractor] etc. ##

**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**

In a recent CI build log for the Rush Stack repo, I noticed this output:

```
2020-08-21T22:43:03.3805464Z 44 of 89: [@microsoft/web-library-build] completed successfully in 11.58 seconds
2020-08-21T22:43:03.3806327Z [@rushstack/module-minifier-plugin] started
2020-08-21T22:43:03.3807673Z >>> @rushstack/module-minifier-plugin
2020-08-21T22:43:03.3808472Z heft test --clean --production
2020-08-21T22:43:03.4326428Z Using local Heft from /home/vsts/work/1/s/webpack/module-minifier-plugin/node_modules/@rushstack/heft
2020-08-21T22:43:03.4335074Z 
2020-08-21T22:43:04.3437086Z Project build folder is "/home/vsts/work/1/s/webpack/module-minifier-plugin"
2020-08-21T22:43:04.3463028Z Starting test
2020-08-21T22:43:04.3564802Z  ---- Clean started ---- 
2020-08-21T22:43:04.3582857Z Deleted 4 paths
2020-08-21T22:43:04.3586604Z  ---- Clean finished (2ms) ---- 
2020-08-21T22:43:04.3662648Z  ---- Compile started ---- 
2020-08-21T22:43:04.4026359Z [copy-static-assets] Copied 0 static assets in 0ms
2020-08-21T22:43:05.0702337Z [typescript] Using TypeScript version 3.5.3
2020-08-21T22:43:05.0703565Z [eslint] Using ESLint version 7.2.0
2020-08-21T22:43:13.5294701Z  ---- Compile finished (9162ms) ---- 
2020-08-21T22:43:13.5306272Z  ---- Bundle started ---- 
2020-08-21T22:43:14.1008070Z [api-extractor] Using API Extractor version 7.9.8
2020-08-21T22:43:15.4903071Z [api-extractor] Analysis will use the bundled TypeScript version 3.9.7
2020-08-21T22:43:17.8259727Z 45 of 89: [@rushstack/localization-plugin] completed successfully in 14.97 seconds
2020-08-21T22:43:17.8261394Z [api-extractor-lib2-test] started
2020-08-21T22:43:18.3640550Z  ---- Bundle finished (4831ms) ---- 
2020-08-21T22:43:18.3642243Z  ---- Test started ---- 
2020-08-21T22:43:18.7891086Z �[0m
2020-08-21T22:43:18.7895344Z Run start. 2 test suites
2020-08-21T22:43:18.7914168Z START src/test/RehydrateAsset.test.ts
2020-08-21T22:43:20.4355031Z PASS src/test/RehydrateAsset.test.ts (duration: ?, 8 passed, 0 failed)
2020-08-21T22:43:20.4362599Z START src/test/MinifiedIdentifier.test.ts
2020-08-21T22:43:20.6307696Z PASS src/test/MinifiedIdentifier.test.ts (duration: ?, 2 passed, 0 failed)
```

The "Bundle started" and "Bundle finished" lines are clearly from the same `module-minifier-plugin` log. But there's Rush output is mixed in the middle of those lines.

**What is the expected behavior?**

The [stream-collator](https://www.npmjs.com/package/@rushstack/stream-collator) is a showcase feature of Rush. It collates log output, enabling developers to view realtime output from parallel builds, without having to awkwardly prefix each log line with an associated project name.

Whereas when the output is interleaved as above, it is difficult to interpret.  Suppose a unit test failed:

```
2020-08-21T22:43:17.8261394Z [api-extractor-lib2-test] started
2020-08-21T22:43:18.3640550Z  ---- Bundle finished (4831ms) ---- 
2020-08-21T22:43:18.3642243Z  ---- Test started ---- 
2020-08-21T22:43:18.7891086Z �[0m
2020-08-21T22:43:18.7895344Z Run start. 2 test suites
2020-08-21T22:43:18.7914168Z START src/test/RehydrateAsset.test.ts
2020-08-21T22:43:18.7914168Z FAIL: Some problem with RehydrateAsset
```

Which project is this referring to?  Because Rush is not collating properly, the above error would seem to apply to `api-extractor-lib2-test`, but actually it is referring to `module-minifier-plugin`.

**Remarks**

We recently switched from `gulp-core-build` to Heft --  maybe Heft is somehow directly writing to the terminal's STDOUT, preventing Rush from capturing and collating the stream?

### PR DESCRIPTION
**Pull request by octogonz**
This is more groundwork for my upcoming PR that will fix issue [2135](https://github.com/microsoft/rushstack/issues/2135)

### CODE REVIEW

File: libraries/node-core-library/src/Terminal/AnsiEscape.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+      return `[${csiCode}]`;
+    });
+
+    if (options.encodeNewlines) {
+      result = result.replace(/\n/g, '[n]').replace(/\r/g, `[r]`);
+    }
+    return result;
+  }
+
+  private static _x(n: number): string | undefined {
```
Comment (iclanton) [inline libraries/node-core-library/src/Terminal/AnsiEscape.ts:62]: Can we give this function a more meaningful name?

File: libraries/node-core-library/src/Terminal/AnsiEscape.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    switch (n) {
```
Comment (iclanton) [inline libraries/node-core-library/src/Terminal/AnsiEscape.ts:63]: Use [this enum](https://github.com/microsoft/rushstack/blob/master/libraries/node-core-library/src/Terminal/Colors.ts#L48-L83)

**Review by iclanton — CHANGES_REQUESTED** [review]

File: libraries/node-core-library/src/Terminal/AnsiEscape.ts (hunk 1 above)
Comment (octogonz) [reply libraries/node-core-library/src/Terminal/AnsiEscape.ts:62]: Haha oops, I forgot to name it 😁

File: libraries/node-core-library/src/Terminal/AnsiEscape.ts (hunk 2 above)
Comment (octogonz) [reply libraries/node-core-library/src/Terminal/AnsiEscape.ts:63]: I started with that, but the strings that it produces are very long, and I found them difficult to read.  Compare:
```
exports[`Colors correctly normalizes color codes for tests 1`] = `"X[BlackForeground]X[DefaultForeground][WhiteForeground]X[DefaultForeground][GrayForeground]
```
versus
```
exports[`Colors correctly normalizes color codes for tests 1`] = `"X[black]X[default][white]X[default][gray]
```

Comment (iclanton) [reply libraries/node-core-library/src/Terminal/AnsiEscape.ts]: No, I meant use that enum instead of the numbers in each `case` statement.

Comment (octogonz) [reply libraries/node-core-library/src/Terminal/AnsiEscape.ts]: Ahhh right -- good idea. :-)  Fixed !

**Review by iclanton — APPROVED** [review]

**Pull request merged**